mobile/src/main/java/ch/epfl/sdp/mobile/ui/play/ExpandableFloatingActionButton.kt

```diff
@@ -0,0 +1,240 @@
+ *
+ * @param onClick the callback which will be called when the button is clicked.
+ * @param icon the slot for the action icon.
+ * @param modifier the [Modifier] for this composable.
+ * @param text the slot for the action text.
+ */
+@Composable
+fun ExpandableFloatingActionButtonItem(
+    onClick: () -> Unit,
+    icon: @Composable () -> Unit,
+    modifier: Modifier = Modifier,
+    text: @Composable () -> Unit,
+) {
+  Row(
+      modifier = modifier.clickable { onClick() }.padding(vertical = 20.dp, horizontal = 24.dp),
+      horizontalArrangement = Arrangement.SpaceBetween,
+      verticalAlignment = Alignment.CenterVertically,
+  ) {
+    // This box is required for intrinsic measurements to work properly, since icons don't quite
+    // support them properly yet.
+    Box(Modifier.size(24.dp), Alignment.Center) { icon() }
+    Spacer(Modifier.width(16.dp))
+    text()
+  }
+}
+
+// Content wrappers with proper Arrangement and Alignment.
+
+@Composable
+private fun CollapsedContent(
+    modifier: Modifier = Modifier,
+    content: @Composable RowScope.() -> Unit,
+) = Row(modifier, Arrangement.Start, Alignment.CenterVertically, content)
+
+@Composable
+private fun ExpandedContent(
+    modifier: Modifier = Modifier,
+    content: @Composable ColumnScope.() -> Unit,
+) = Column(modifier.width(IntrinsicSize.Max), Arrangement.Top, Alignment.Start, content)
+
+// State management.
+
+/**
+ * Creates and remembers an [ExpandableFloatingActionButtonState].
+ *
+ * @param initiallyExpanded true if the floating action button should initially be expanded.
+ */
+@Composable
+fun rememberExpandableFloatingActionButtonState(
+    initiallyExpanded: Boolean = false,
+): ExpandableFloatingActionButtonState {
+  val expanded = rememberSaveable { mutableStateOf(initiallyExpanded) }
+  return SnapshotExpandableFloatingActionButtonState(expanded)
+}
+
+/**
+ * An implementation of an [ExpandableFloatingActionButtonState] that delegates to a backing
+ * [MutableState].
+ *
+ * @param state the [MutableState] to which the expanded state is delegated.
+ */
+private class SnapshotExpandableFloatingActionButtonState(
+    state: MutableState<Boolean>,
+) : ExpandableFloatingActionButtonState {
+  override var expanded by state
+}
+
+// Configuration.
+
+/**
+ * The default [ExpandableFloatingActionButtonColors], which default to some branded colors from the
+ * Pawnies theme.
+ */
+private object DefaultExpandableFloatingActionButtonColors : ExpandableFloatingActionButtonColors {
+
+  @Composable
+  override fun background(expanded: Boolean): State<Color> {
+    return animateColorAsState(if (expanded) Beige050 else Green800)
+  }
+
+  @Composable
+  override fun border(expanded: Boolean): State<Color> {
+    return animateColorAsState(if (expanded) Green100 else Green100.copy(alpha = 0f))
+  }
+
+  @Composable
+  override fun contentColor(expanded: Boolean): State<Color> {
+    return animateColorAsState(if (expanded) Green800 else Green100)
```

Comment:
[Question] Does 'animateColorAsState' make the `fun contentColor(expanded: Boolean)` returns a slightly different color tint starting from `Green800 `down to `Green100 ` each time the composable gets updated ? Meaning is the returned State<Color> responsible for the triggered updates ?

Reply:
An important concept of compose is that `State` values which are read by the composition / the layout phase / the drawing phase and which are updated will trigger an invalidation of the composition (respectively layout / drawing phases) if they're updated.

`animateColorAsState` will create a `State` whose value gets smoothly updated at 60 fps when the target changes. **The `State` updates therefore drive the recompositions, relayout or redraws**, depending on how the read value is used.

This also means that once the animation is done (the color is done animating), the `State` does not get updated with new values, and there are no more recompositions.